documents/en/settings.md
Settings
===
[back](../../README.md)

Table of Contents

<!--ts-->
* [Settings](#settings)
   * [Introduction](#introduction)
   * [Environment variables called at compile time (config.nims)](#environment-variables-called-at-compile-time-confignims)
      * [DB_SQLITE :string = "true"](#db_sqlite-string--true)
      * [DB_POSTGRES :string = "false"](#db_postgres-string--false)
      * [DB_MYSQL :string = "false"](#db_mysql-string--false)
      * [DB_MARIADB :string = "false"](#db_mariadb-string--false)
      * [SESSION_TYPE :string = "file"](#session_type-string--file)
      * [USE_LIBSASS :string = "false"](#use_libsass-string--false)
   * [Environment variables called at runtime (.env)](#environment-variables-called-at-runtime-env)
      * [SECRET_KEY :string](#secret_key-string)
      * [DB_DATABASE :string = ""](#db_database-string--)
      * [DB_USER :string = ""](#db_user-string--)
      * [DB_PASSWORD :string = ""](#db_password-string--)
      * [DB_HOST :string = "sqlite"](#db_host-string--sqlite)
      * [DB_PORT :int = 5432](#db_port-int--5432)
      * [DB_MAX_CONNECTION :int = 1](#db_max_connection-int--1)
      * [SESSION_DB_PATH :string = getCurrentDir() / "session.db"](#session_db_path-string--getcurrentdir--sessiondb)
      * [COOKIE_DOMAINS :string = ""](#cookie_domains-string--)
   * [Settings configured in the Settings object](#settings-configured-in-the-settings-object)
      * [HOST :string = "0.0.0.0"](#host-string--0000)
      * [PORT :int = 8000](#port-int--8000)
      * [LOG_TO_CONSOLE :bool = true](#log_to_console-bool--true)
      * [LOG_TO_FILE :bool = true](#log_to_file-bool--true)
      * [ERROR_LOG_TO_FILE :bool = true](#error_log_to_file-bool--true)
      * [LOG_DIR :string = getCurrentDir() / "logs"](#log_dir-string--getcurrentdir--logs)
      * [SESSION_TIME :int = 120](#session_time-int--120)
      * [SESSION_EXPIRE_ON_CLOSE: bool = false](#session_expire_on_close-bool--false)
      * [LOCALE :string = "en"](#locale-string--en)

<!-- Created by https://github.com/ekalinin/github-markdown-toc -->
<!-- Added by: root, at: Sat Jun 22 11:26:30 UTC 2024 -->

<!--te-->

## Introduction
Basolato settings are defined in three ways:

- Written in `config.nims`, managed by git, and called at compile time
- Defined in `.env` during development or in server environment variables during production, not managed by git, and called at runtime
- Recorded in an instance of `Settings.new()`, managed by git, and called at runtime

## Environment variables called at compile time (config.nims)
Changes require recompilation to take effect.

### DB_SQLITE :string = "true"
Whether to connect to Sqlite.

### DB_POSTGRES :string = "false"
Whether to connect to PostgreSQL.

### DB_MYSQL :string = "false"
Whether to connect to MySQL.

### DB_MARIADB :string = "false"
Whether to connect to MariaDB.

### SESSION_TYPE :string = "file"
Type of session DB. Either file or redis.

### USE_LIBSASS :string = "false"
Whether to enable libsaas and allow the use of SASS/SCSS in views.

Sample config.nims
```nim
nimCopyputEnv("DB_SQLITE", $true) # "true" or "false"
# putEnv("DB_POSTGRES", $true) # "true" or "false"
# putEnv("DB_MYSQL", $true) # "true" or "false"
# putEnv("DB_MARIADB", $true) # "true" or "false"
putEnv("SESSION_TYPE", "file") # "file" or "redis"
putEnv("USE_LIBSASS", $false) # "true" or "false"
```

## Environment variables called at runtime (.env)
Restart the application to apply changes.

### SECRET_KEY :string
A 100-character key used for encrypting session IDs, etc.

### DB_DATABASE :string = ""
Database name of the target RDB.

### DB_USER :string = ""
Username for connecting to the RDB.

### DB_PASSWORD :string = ""
Password for connecting to the RDB.

### DB_HOST :string = "sqlite"
Host of the target RDB. Specify the absolute file path for Sqlite, or the `host` for MySQL and PostgreSQL.

### DB_PORT :int = 5432
Location of the target RDB. Leave empty for Sqlite, specify the `port number` for MySQL and PostgreSQL.

### DB_MAX_CONNECTION :int = 1
Number of connection pools created for asynchronous RDB connections.  
For multi-threaded applications, set this to "maximum connections / number of threads".

### SESSION_DB_PATH :string = getCurrentDir() / "session.db"
Location of the session DB.
Set the absolute file path for file sessions, or `host:port` for Redis.

### COOKIE_DOMAINS :string = ""
Set the target domains for issuing cookies.


Sample .env
```sh
# Secret
SECRET_KEY="GRzV3jfgN8BgFhtiyoLV1aTNE6Evh9r1GLkpBpUCpioXy6ifo10fEL846MTRrd3cpOHMKsYCs1hNQDDYJ3NEOs2mEPYTemU3iGnm"

# DB Connection
DB_DATABASE="database" # sqlite file path or database name
DB_USER="user"
DB_PASSWORD="password"
DB_HOST="127.0.0.1"  # host ip address
DB_PORT=5432 # postgres default...5432, mysql default...3306
DB_MAX_CONNECTION=95 # should be smaller than (DB max connection / running num processes)
DB_TIMEOUT=30 # seconds

# Session db
# Session type, file or redis, is defined in config.nims
SESSION_DB_PATH="./session.db" # Session file path or redis host:port. ex:"127.0.0.1:6379"

COOKIE_DOMAINS="" # to specify multiple domains, "sample.com, sample.org"
```

## Settings configured in the Settings object
Non-sensitive information that should be set at runtime but not in environment variables is defined in `Settings.new()`.

### HOST :string = "0.0.0.0"
Hostname of the server to start.

### PORT :int = 8000
Port number of the server to start.

### LOG_TO_CONSOLE :bool = true
Set to true to display logs in the terminal, false otherwise.

### LOG_TO_FILE :bool = true
Set to true to output logs to a file, false otherwise.

### ERROR_LOG_TO_FILE :bool = true
Set to true to output error logs to a file, false otherwise.

### LOG_DIR :string = getCurrentDir() / "logs"
Absolute path of the log output directory.

### SESSION_TIME :int = 120
Set the session timeout period in minutes.

### SESSION_EXPIRE_ON_CLOSE: bool = false
Set to true to automatically delete the session when the browser is closed.

### LOCALE :string = "en"
Language for displaying validation messages.

|Language|Environment Variable|
|---|---|
|English|en|
|Japanese|ja|

Sample code main.nim
```nim
import basolato

let routs = @[
  Route.get("/", example_controller.index)
]

let settings = Settings.new(
  host = "127.0.0.1",
  port = 8000,
  # Logging
  logToConsole = true,
  logToFile = false,
  errorLogToFile = false,
  logDir = "./logs",
  # Session db
  sessionTime = 120, # default 120, minutes of 2 hours
  sessionExpireOnClose = false,
  # other
  locale = "en",
)

serve(routes, settings)
```


documents/ja/settings.md
設定
===
[戻る](../../README.md)

目次
<!--ts-->
* [設定](#設定)
   * [イントロダクション](#イントロダクション)
   * [コンパイル時に呼ばれる環境変数 (config.nims)](#コンパイル時に呼ばれる環境変数-confignims)
      * [DB_SQLITE :string = "true"](#db_sqlite-string--true)
      * [DB_POSTGRES :string = "false"](#db_postgres-string--false)
      * [DB_MYSQL :string = "false"](#db_mysql-string--false)
      * [DB_MARIADB :string = "false"](#db_mariadb-string--false)
      * [SESSION_TYPE :string = "file"](#session_type-string--file)
      * [USE_LIBSASS :string = "false"](#use_libsass-string--false)
   * [実行時に呼ばれる環境変数 (.env)](#実行時に呼ばれる環境変数-env)
      * [SECRET_KEY :string](#secret_key-string)
      * [DB_DATABASE :string](#db_database-string)
      * [DB_USER :string](#db_user-string)
      * [DB_PASSWORD :string](#db_password-string)
      * [DB_HOST :string](#db_host-string)
      * [DB_PORT :int](#db_port-int)
      * [DB_MAX_CONNECTION :int](#db_max_connection-int)
      * [SESSION_DB_PATH :string](#session_db_path-string)
      * [COOKIE_DOMAINS :string](#cookie_domains-string)
   * [Settingオブジェクトで設定するもの](#settingオブジェクトで設定するもの)
      * [HOST :string = "0.0.0.0"](#host-string--0000)
      * [PORT :int = 8000](#port-int--8000)
      * [LOG_TO_CONSOLE :bool = true](#log_to_console-bool--true)
      * [LOG_TO_FILE :bool = true](#log_to_file-bool--true)
      * [ERROR_LOG_TO_FILE :bool = true](#error_log_to_file-bool--true)
      * [LOG_DIR :string = getCurrentDir() / "logs"](#log_dir-string--getcurrentdir--logs)
      * [SESSION_TIME :int = 120](#session_time-int--120)
      * [SESSION_EXPIRE_ON_CLOSE: bool = false](#session_expire_on_close-bool--false)
      * [LOCALE :string = "en"](#locale-string--en)

<!-- Created by https://github.com/ekalinin/github-markdown-toc -->
<!-- Added by: root, at: Sat Jun 22 11:26:16 UTC 2024 -->

<!--te-->

## イントロダクション
Basolatoの設定は、3つの方法で定義されます。
- `config.nims`に書かれ、git管理され、コンパイル時に呼ばれるもの
- 開発時は`.env`に、本番稼働時はサーバーの環境変数に定義され、git管理されず、実行時に呼ばれるもの
- `Settings.new()`のインスタンスに記録され、git管理され、実行時に呼ばれるもの


## コンパイル時に呼ばれる環境変数 (config.nims)
変更を適用するには再度コンパイルする必要があります。

### DB_SQLITE :string = "true"
`Sqlite`に接続するかどうかです。

### DB_POSTGRES :string = "false"
`PostgreSQL`に接続するかどうかです。

### DB_MYSQL :string = "false"
`MySQL`に接続するかどうかです。

### DB_MARIADB :string = "false"
`MariaDB`に接続するかどうかです。

### SESSION_TYPE :string = "file"
セッションDBの種類です。`file`、`redis`のいずれかになります。

### USE_LIBSASS :string = "false"
`libsaas`を有効にし、ビューの中でSASS/SCSSを使えるようにするかどうかです。

サンプル config.nims
```nim
putEnv("DB_SQLITE", $true) # "true" or "false"
# putEnv("DB_POSTGRES", $true) # "true" or "false"
# putEnv("DB_MYSQL", $true) # "true" or "false"
# putEnv("DB_MARIADB", $true) # "true" or "false"
putEnv("SESSION_TYPE", "file") # "file" or "redis"
putEnv("USE_LIBSASS", $false) # "true" or "false"
```

## 実行時に呼ばれる環境変数 (.env)
変更を適用するには再度アプリケーションを起動してください。

### SECRET_KEY :string
セッションIDなどの暗号化に使われる100文字のキーです。

### DB_DATABASE :string
接続先RDBのデータベース名です。

### DB_USER :string
RDBに接続するためのユーザー名です。

### DB_PASSWORD :string
RDBに接続するためのパスワードです。

### DB_HOST :string
接続先RDBのホストです。Sqliteを使う場合にはファイルの絶対パス、MySQL, PostgreSQLを使う場合には`ホスト`を指定してください。

### DB_PORT :int
接続先RDBの場所です。Sqliteを使う場合には空、MySQL, PostgreSQLを使う場合には`ポート番号`を指定してください。

### DB_MAX_CONNECTION :int
RDBに非同期接続する時に作るコネクションプールの数です。  
アプリケーションをマルチスレッドで動かす場合には、「接続可能数 / スレッド数」になるようにしてください。

### SESSION_DB_PATH :string
接続先セッションDBの場所です。  
ファイルセッションを使う場合にはファイルの絶対パスを、Redisを使う場合には`ホスト:ポート`を設定してください。

### COOKIE_DOMAINS :string
クッキーを発行する対象ドメインを設定してください。

サンプル .env
```sh
# Secret
SECRET_KEY="GRzV3jfgN8BgFhtiyoLV1aTNE6Evh9r1GLkpBpUCpioXy6ifo10fEL846MTRrd3cpOHMKsYCs1hNQDDYJ3NEOs2mEPYTemU3iGnm"

# DB Connection
DB_DATABASE="database" # sqlite file path or database name
DB_USER="user"
DB_PASSWORD="password"
DB_HOST="127.0.0.1"  # host ip address
DB_PORT=5432 # postgres default...5432, mysql default...3306
DB_MAX_CONNECTION=95 # should be smaller than (DB max connection / running num processes)
DB_TIMEOUT=30 # secounds

# Session db
# Session type, file or redis, is defined in config.nims
SESSION_DB_PATH="./session.db" # Session file path or redis host:port. ex:"127.0.0.1:6379"

COOKIE_DOMAINS="" # to specify multiple domains, "sample.com, sample.org"
```

## Settingオブジェクトで設定するもの
実行時に呼ばれ、環境変数に設定すべき値のような機密情報でないものは、`Settings.new()`で定義します。

### HOST :string = "0.0.0.0"
起動するサーバーのホスト名です。

### PORT :int = 8000
起動するサーバーのポート番号です。

### LOG_TO_CONSOLE :bool = true
ログをターミナルに表示するなら`true`を、しないなら`false`を設定してください。

### LOG_TO_FILE :bool = true
ログをファイル出力するなら`true`を、しないなら`false`を設定してください。

### ERROR_LOG_TO_FILE :bool = true
エラーログをファイル出力するなら`true`を、しないなら`false`を設定してください。

### LOG_DIR :string = getCurrentDir() / "logs"
ログ出力先ディレクトリの絶対パスです。

### SESSION_TIME :int = 120
セッションタイムアウトになる期限を分で設定してください。

### SESSION_EXPIRE_ON_CLOSE: bool = false
ブラウザを閉じた時にセッションを自動的に削除する場合はtrueを設定してください。

### LOCALE :string = "en"
バリデーションメッセージを表示する言語です。

|言語|環境変数|
|---|---|
|英語|en|
|日本語|ja|


サンプルコード main.nim
```nim
import basolato

let routs = @[
  Route.get("/", example_controller.index)
]

let settings = Settings.new(
  host = "127.0.0.1",
  port = 8000,
  # Logging
  logToConsole = true,
  logToFile = false,
  errorLogToFile = false,
  logDir = "./logs",
  # Session db
  sessionTime = 120, # default 120, minutes of 2 hours
  sessionExpireOnClose = false,
  # other
  locale = "en",
)

serve(routes, settings)
```


src/basolato/cli/functions/make/config.nim
import std/os
import std/strformat
import std/terminal
import ./utils


template createConfigCommon(target, CONFIG:untyped) =
  block:
    let targetPath = getCurrentDir() / target
    if not isFileExists(targetPath):
      let f = open(targetPath, fmWrite)
      defer: f.close()
      f.write(CONFIG)

      let message = "Created " & targetPath
      styledWriteLine(stdout, fgGreen, bgDefault, message, resetStyle)


let currentDir = getCurrentDir()

proc makeConfig*():int =
  createConfigCommon("config.nims"):
    &"""
import std/os

putEnv("DB_SQLITE", $true) # "true" or "false"
# putEnv("DB_POSTGRES", $true) # "true" or "false"
# putEnv("DB_MYSQL", $true) # "true" or "false"
# putEnv("DB_MARIADB", $true) # "true" or "false"
putEnv("SESSION_TYPE", "file") # "file" or "redis"
putEnv("LIBSASS", $false) # "true" or "false"
"""

  createConfigCommon(".env"):
    &"""
# Secret
SECRET_KEY="{randStr(100)}"

# DB Connection
DB_DATABASE="{currentDir}/db.sqlite3" # sqlite file path or database name
DB_USER=""
DB_PASSWORD=""
DB_HOST=""  # host ip address
DB_PORT=0 # postgres default...5432, mysql default...3306
DB_MAX_CONNECTION=95 # should be smaller than (DB max connection / running num processes)
DB_TIMEOUT=30 # secounds

# Session db
# Session type, file or redis, is defined in config.nims
SESSION_DB_PATH="{currentDir}/session.db" # Session file path or redis host:port. ex:"127.0.0.1:6379"

COOKIE_DOMAINS="" # to specify multiple domains, "sample.com, sample.org"
"""

  createConfigCommon(".env.example"):
    &"""
# Secret
SECRET_KEY=""

# DB Connection
DB_DATABASE="" # sqlite file path or database name
DB_USER=""
DB_PASSWORD=""
DB_HOST=""  # host ip address
DB_PORT=0 # postgres default...5432, mysql default...3306
DB_MAX_CONNECTION=95 # should be smaller than (DB max connection / running num processes)
DB_TIMEOUT=30 # secounds

# Session db
# Session type, file or redis, is defined in config.nims
SESSION_DB_PATH="" # Session file path or redis host:port. ex:"127.0.0.1:6379"

COOKIE_DOMAINS="" # to specify multiple domains, "sample.com, sample.org"
"""
